toponetx/classes/path.py

```diff
@@ -0,0 +1,227 @@
+    from the elementary p-path, which is identical to a simplex.
+
+    References
+    ----------
+    .. [1] Truong and Chin.
+        Generalizing Topological Graph Neural Networks with Paths.
+        https://arxiv.org/abs/2308.06838.pdf
+    .. [2] Grigor'yan, Lin, Muranov, and Yau.
+        Homologies of path complexes and digraphs.
+        https://arxiv.org/pdf/1207.2834.pdf
+
+    Examples
+    --------
+    >>> path1 = Path((1, 2, 3))
+    >>> list(path1.boundary)
+    []
+    >>> path2 = Path((1, 2, 3), construct_boundaries=True)
+    >>> list(path2.boundary)
+    [(2, 3), (1, 3), (1, 2)]
+    >>> path3 = Path((1, 2, 3), construct_boundaries=True, allowed_paths=[(1, 2), (2, 3)])
+    >>> list(path3.boundary)
+    [(2, 3), (1, 2)]
+    >>> path4 = Path((1, 2, 3), construct_boundaries=True, allowed_paths=[(1, 3), (2, 3)])
+    >>> list(path4.boundary)
+    [(2, 3), (1, 3)]
+    """
+
+    def __init__(
+        self,
+        elements: Sequence[Hashable],
+        name: str = "",
+        construct_boundaries: bool = False,
+        reserve_sequence_order: bool = False,
+        allowed_paths: Iterable[tuple[Hashable]] | None = None,
+        **attr,
+    ) -> None:
+        self.__check_inputs(elements, reserve_sequence_order)
+        if isinstance(elements, int) or isinstance(elements, str):
+            elements = [elements]
+        super().__init__(tuple(elements), name, **attr)
+        if len(set(elements)) != len(self.elements):
+            raise ValueError("A p-path cannot contain duplicate nodes.")
+
+        self.construct_boundaries = construct_boundaries
+        self.reserve_sequence_order = reserve_sequence_order
+        if construct_boundaries:
+            self._boundaries = self.construct_path_boundaries(
+                elements,
+                reserve_sequence_order=reserve_sequence_order,
+                allowed_paths=allowed_paths,
+            )
+        else:
+            self._boundaries = list()
+
+    @staticmethod
+    def construct_path_boundaries(
+        elements: Sequence[Hashable],
+        reserve_sequence_order: bool = False,
+        allowed_paths: Iterable[tuple[Hashable]] | None = None,
+    ) -> list[tuple[Hashable]]:
+        """
+        Return list of elementary p-path objects representing the boundaries.
+
+        Parameters
+        ----------
+        elements: Sequence[Hashable]
+            The nodes in the elementary p-path.
+        reserve_sequence_order : bool, default=False
+            If True, reserve the order of the sub-sequence of nodes in the elementary p-path.
+            Else, the sub-sequence of nodes in the elementary p-path will be reversed if the first index is larger than the last index.
+        allowed_paths : Iterable[tuple[Hashable]], optional
+            A list of allowed boundaries. If None, all possible boundaries are constructed (similarly to simplex).
+
+        Returns
+        -------
+        boundaries : list[tuple[Hashable]]
```

Comment:
Single return values do not have a name.

```suggestion
        list[tuple[Hashable]]
```